core/remotes/docker/config/hosts_resolver_test.go

```diff
@@ -0,0 +1,318 @@
+		})
+		tokenServer, tokenCert := newTLSServer(tokenHandler)
+		tokenBase := tokenServer.URL + "/token"
+
+		// Wrap with token auth
+		tokenWrapped := http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			auth := strings.ToLower(r.Header.Get("Authorization"))
+			if auth != "bearer perfectlyvalidopaquetoken" {
+				authHeader := fmt.Sprintf("Bearer realm=%q,service=registry,scope=\"repository:testname:pull,pull\"", tokenBase)
+				if strings.HasPrefix(auth, "bearer ") {
+					authHeader = authHeader + ",error=" + auth[7:]
+				}
+				rw.Header().Set("WWW-Authenticate", authHeader)
+				rw.WriteHeader(http.StatusUnauthorized)
+				return
+			}
+			h.ServeHTTP(rw, r)
+		})
+
+		upstream, upstreamCert := newTLSServer(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			t.Fatal("upstream server shouldn't be called, given server is specified in hosts.toml")
+		}))
+		base := upstream.URL[8:] // strip "https://"
+		mirror, mirrorCert := newTLSServer(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			mirrorTriggered.Store(true)
+			mirrorOnlyHeader(t, r)
+			tokenWrapped.ServeHTTP(rw, r)
+		}))
+		server, serverCert := newTLSServer(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			serverTriggered.Store(true)
+			serverOnlyHeader(t, r)
+		}))
+
+		capool := x509.NewCertPool()
+		for _, cert := range []*x509.Certificate{tokenCert, upstreamCert, mirrorCert, serverCert} {
+			capool.AddCert(cert)
+		}
+
+		hostTOML := fmt.Sprintf(hostTOMLTemplate, server.URL, mirror.URL)
+		var hostDir, hostFilePath string
+		if runtime.GOOS == "windows" {
+			hostDir = filepath.Join(dir, strings.ReplaceAll(hostDirectory(base), ":", ""))
+		} else {
+			hostDir = filepath.Join(dir, hostDirectory(base))
+		}
+		hostFilePath = filepath.Join(hostDir, "hosts.toml")
+
+		err := os.MkdirAll(hostDir, 0755)
+		assert.NoError(t, err)
+		err = os.WriteFile(hostFilePath, []byte(hostTOML), 0644)
+		assert.NoError(t, err)
+
+		hostOptions := HostOptions{
+			HostDir: HostDirFromRoot(dir),
+			DefaultTLS: &tls.Config{
+				RootCAs: capool,
+			},
+		}
+
+		options := docker.ResolverOptions{
+			Hosts: ConfigureHosts(context.TODO(), hostOptions),
+		}
+
+		if !enableMirror {
+			mirror.Close()
+		}
+		if !enableServer {
+			server.Close()
+		}
+		closer := func() {
+			tokenServer.Close()
+			upstream.Close()
+			if enableMirror {
+				mirror.Close()
+			}
+			if enableServer {
+				server.Close()
+			}
+		}
+
+		return base, options, closer
+	}
+
+	runBasicTest(t, name, testFn)
+
+	switch {
+	case enableMirror:
+		assert.True(t, mirrorTriggered.Load())
+		assert.False(t, serverTriggered.Load())
+	case enableServer:
+		assert.False(t, mirrorTriggered.Load())
+		assert.True(t, serverTriggered.Load())
+	default:
+		assert.False(t, mirrorTriggered.Load())
+		assert.False(t, serverTriggered.Load())
+	}
+}
+
+func runBasicTest(t *testing.T, name string, sf func(h http.Handler) (string, docker.ResolverOptions, func())) {
```

Comment:
This is largely copied from resolver_test.go to avoid circular imports. 

https://github.com/containerd/containerd/blob/2c042ae0de7b9d69df8df8e370e2d802f10f9ee5/core/remotes/docker/resolver_test.go#L773